### ISSUE TITLE
Hook directly after jobs.lua is loaded

### ISSUE THREAD

**Issue by Partixel** [FPtje/DarkRP#1871]
Need the hook so that my addon can convert the jobs and as such set the global variables TEAM_* before shipments etc are loaded.

**Comment by Partixel** [comment]
Also grats on 3400+ commits.

**Comment by FPtje** [comment]
The TEAM_ vars are set to numbers, why not try to keep them intact?

The problem I have here is that there are hooks for before loading, during
loading and after loading. Now you want to have a hook for basically every
step along the way? That's too bloated and it shouldn't be done unless
there really is no other way.

**Comment by Partixel** [comment]
Fair point, I'll keep trying. The problem is at the point where my addon converts the default jobs.lua etc to my addons format which causes the TEAM_ vars to lose their place

**Comment by FPtje** [comment]
That only happens when you make new tables and or change their positions in
RPExtraTeams

**Comment by Partixel** [comment]
Let me rephrase that:

For the shipments.lua to work with my addon, the TEAM_ vars need to be the number that my addon registers them as,  but my addon registers them AFTER the shipments.lua is ran so they are nil

**Comment by Neosinos** [comment]
Wouldnt it be better to store the Teams as unique Text Values (eg
TEAM_MAYOR) instead of numbers which change when editing/sorting the Jobs?

Am Dienstag, 31. März 2015 schrieb Partixel :

**Comment by Partixel** [comment]
Because ( Without being sloppy ) I can't get the current name the user uses for their TEAM_ so I create a new one based on the teams command so the one loaded in jobs.lua could be different if they don't use TEAM_COMMAND

**Comment by Partixel** [comment]
Was that clear enough? It's kinda hard for me to explain what I mean.

**Comment by FPtje** [comment]
Why do you need the TEAM_ name? You have the values, they're the keys of
the RPExtraTeams table.

Your ui using them is not really a good reason. A job picker would work
much better than people having to enter the TEAM_ vars manually.

**Comment by Partixel** [comment]
Here, I'll give an example:

In jobs.lua which is loaded first:

		TEAM_HOBOL = DarkRP.createJob("Hobo Leader", {
			command = "hoboleader"
		})

My addon would then interpret that into the TEAM_COMMAND format:

		TEAM_HOBOLEADER = DarkRP.createJob("Hobo Leader", {
			command = "hoboleader"
		})

This will mean when they use TEAM_HOBOLEADER in the shipments.lua it will be nil because my addon is loaded on the loadCustomDarkRPItems which is after the shipments.lua is ran

**Comment by Partixel** [comment]
It would be fine if they initially have the variable named in format my addon uses ( TEAM_COMMAND ) because it would be the teams number at the point shipments.lua is ran

**Comment by FPtje** [comment]
You want to create a NEW global variable so that they can use it in the
shipments? Why not let them use the TEAM_HOBOL they already had? That's
what they're going to put in the shipments, not whatever your addon thinks
of!

**Comment by FPtje** [comment]
Besides, you're going to break a LOT more than just the team numbers
themselves if you change the numbers of jobs. That's really asking for
trouble.

**Comment by Partixel** [comment]
The numbers of jobs isn't change, only the name of the global variable because I can't get the current name they use in any non-sloppy manor

**Comment by Partixel** [comment]
It's hard to explain :(

**Comment by Partixel** [comment]
I'll just go cry in a corner until I learn to code better because I'm a noob.

Cya all!

**Comment by FPtje** [comment]
You were never supposed to use the TEAM_ variables for manipulation of
jobs. If you look in DarkRP itself, the TEAM_ variables are used nowhere
except for the default items. Surely if DarkRP doesn't need it to do
everything it does with the jobs, you don't either.

Custom shipments and entities don't even have global variables. Somehow
that works fine too. They get in the right places in the F4 menu, all the
commands are created etc. etc.

The TEAM_ variables are for cross reference between the custom items, not
for DarkRP itself or anything else to use.

2015-03-31 20:26 GMT+02:00 Partixel <notifications@github.com>:

**Comment by Partixel** [comment]
Yeah, that's what I meant. Shipments use the TEAM_ variables ( or should I say, they use the values ) for the allowed property:

allowed = {TEAM_GUN, TEAM_MEDIC}, -- OPTIONAL, which teams are allowed to buy this shipment/separate gun

**Comment by Partixel** [comment]
I was trying to make the jobs that my addon creates accessible within the shipments.lua for the allowed table

**Comment by Partixel** [comment]
If I could explain it in a non-retarded way I would do, but, unfortunately, I'm too retarded to do so. Sorry :(

**Comment by Partixel** [comment]
I'll give it one more go;

My addon sets the global variables for the jobs TEAM_ because the user may want to make a job be allowed to sell a shipment. If the job isn't in the jobs.lua but is added entirely by my addon then it would be nil at the point shipments.lua is ran.

**Comment by FPtje** [comment]
You could load the jobs before everything else in
http://wiki.darkrp.com/index.php/Hooks/Shared/DarkRPStartedLoading

That or you could do EVERYTHING after loading and tell the user "it's
either the ui or the darkrpmod files". That would mean you discard the
custom items from the darkrpmod and load the ones modified/created by your
ui instead. It would be easier for you to remain compatible. It would ALSO
be easier in the long run for the end user. Think of them asking this
question:

"I can edit jobs through the jobs file or through the ui. Which is better?
faster? which is more stable?". These questions might not make sense to
you, but end users really do ask them.

In the version I got from you, you load the initial jobs from the custom
files, but after a changelevel those jobs were already ignored. You could
go down that road and demand that everything is done through the ui.

2015-03-31 20:54 GMT+02:00 Partixel <notifications@github.com>:

**Comment by Partixel** [comment]
Yeah, I'm planning on making it so you can edit shipments etc from the ui but I was hoping in the meantime this may suffice.

**Comment by Partixel** [comment]
How would I get GAMEMODE / DarkRP in the DarkRPStartedLoading hook?

**Comment by Partixel** [comment]
I tried using the DarkRPFinishedLoading hook at one point, but the same hook is used to load the jobs and is ran before my hook is ran

**Comment by FPtje** [comment]
Oh, you don't. None of the functions are there yet.

**Comment by FPtje** [comment]
loadCustomDarkRPItems loads after all jobs and stuff are done

**Comment by Partixel** [comment]
Yeah, that's what I thought. Is there any way to force my hook.Add to run before the modifcationloaders hook.Add? ( for the DarkRPFinishedLoading hook )

**Comment by FPtje** [comment]
There will be soon.

**Comment by Partixel** [comment]
Sweet, that'll do... that'll do pig..

I can check if my addon has converted their default jobs and if it has then it runs it that hook, else it runs after the jobs.lua etc is added. That _should_ fix my issue.

**Comment by FPtje** [comment]
pig?

2015-03-31 21:07 GMT+02:00 Partixel <notifications@github.com>:

**Comment by Partixel** [comment]
It comes from some movie

**Comment by Partixel** [comment]
https://www.youtube.com/watch?v=rjQtzV9IZ0Q

**Comment by FPtje** [comment]
Holy shit, fuckin Babe. I never expected *that* on my Github.
Thanks for this particular enrichment.

2015-03-31 21:09 GMT+02:00 Partixel <notifications@github.com>:

**Comment by Partixel** [comment]
No problem!

**Comment by CodedNil** [comment]
You guys.. Are weird.

**Comment by Partixel** [comment]
Aw crap, now I need a way to prevent jobs.lua from being loaded. Any possibility of remove function for the "customFiles" table in modifcationloader.lua ?

**Comment by FPtje** [comment]
Can't you return a value in the hook or something?

**Comment by Partixel** [comment]
I don't believe that would prevent the gamemodes hook from running. That would also prevent more then just the jobs.lua from running.

**Comment by FPtje** [comment]
Yes, yes it would.

**Comment by Partixel** [comment]
Also, merry christmass!

**Comment by FPtje** [comment]
Anal cavity

2015-04-01 12:10 GMT+02:00 Partixel <notifications@github.com>:

**Comment by Partixel** [comment]
Nostril sex

**Comment by FPtje** [comment]
Alright, that's enough

2015-04-01 12:27 GMT+02:00 Partixel <notifications@github.com>:

**Comment by Partixel** [comment]
Okay. Anyway, as I was saying, a way to specifically disable the loading of a custom item would be great.

**Comment by FPtje** [comment]
You already can. You'd just have to include the other files manually.

2015-04-01 12:36 GMT+02:00 Partixel <notifications@github.com>:

**Comment by Partixel** [comment]
Well okay then, now we're getting somewhere! -goes off to cry in a corner-

**Comment by Partixel** [comment]
Returning prevents modules and languages from loading too... If only there were, say, a function to remove objects from the customFiles table... *sigh*

**Comment by FPtje** [comment]
Yes, like I said, you have to include them manually.

Jesus christ, what you're doing is already retarded. I'm not making DarkRP
specifically for you. You can do what you need to do well enough. It has to
end fucking somewhere.

2015-04-01 13:29 GMT+02:00 Partixel <notifications@github.com>:

**Comment by Partixel** [comment]
All of those local functions that I have to copy over... sigh

I'm only messing with you, don't worry.

**Comment by FPtje** [comment]
I already told you it's probably better to do everything after the loading and make the ui override the custom files.

**Comment by Partixel** [comment]
Yeah, I am doing. Until I finish off the shipment editor and such I'll just copy the modloader and temporarily remove certain files from loading.

**Comment by Partixel** [comment]
Finally got it working, cheers for the help babe!
